File: 268-MissingNumber/268-MissingNumber.java
// Last updated: 8/16/2026, 9:50:44 PM
class Solution {
    public int missingNumber(int[] nums) {
        // int n = nums.length;
        // Arrays.sort(nums);
        // for(int i = 0; i < n ; i ++) {
        //     if(nums[i] != i) return i;
        // }
        // return n;

        // // method 2 - hashSet
        // int n = nums.length;
        // Set<Integer> set = new HashSet<>();
        // for(int num: nums) set.add(num);
        // for(int i = 0; i <= n; i ++) {
        //     if(!set.contains(i)) return i;
        // }
        // return -1;

        // // method 3 - Math
        // // (1+2+3) - (1+2) = 3 (missing number)
        // int n = nums.length;
        // int sum1 = (int) (1+n)*n/2;
        // int sum2 = 0;
        // for(int num: nums) sum2 += num;
        // return sum1 - sum2;

        // method 4 - bit operation - xor
        // a^b^b = a
        // xor of index [0~n] and nums[i]
        // a^c^b^d^b^a^d = a^a^b^b^d^d^c = c (the missing one)
        int xor = 0;
        int n = nums.length;
        for(int i = 0; i < n; i ++) {
            xor = xor ^ i ^ nums[i];
        }
        int res = xor ^ n;
        return res;
    }
}

File: 268-MissingNumber/268-MissingNumber.py
# Last updated: 8/31/2026, 10:48:16 PM
1class Solution:
2    def missingNumber(self, nums: List[int]) -> int:
3        # # XOR: nums ^ range(n) -> even occurence = 0, odd occurence = 1 -> the missing one
4
5        # n = len(nums)
6        # res = n     # put n in res first, as nums index [0, n-1]
7
8        # # we want to achieve: 0^1^2^...^(n-1)^n ^ nums[0]^nums[1]^...^nums[n-1]
9        # for i, num in enumerate(nums):
10        #     res ^= i ^ num
11        
12        # return res
13
14
15        # sum: sum(0,1,2,...n) - sum(nums) = missing
16        # target_sum = sum(list(range(len(nums) + 1)))
17        n = len(nums)
18        target_sum = (n * (n + 1)) // 2
19
20        return target_sum - sum(nums)